tianshou/data/utils/converter.py

```diff
@@ -80,3 +82,68 @@ def to_torch_as(
     """
     assert isinstance(y, torch.Tensor)
     return to_torch(x, dtype=y.dtype, device=y.device)
+
+
+# Note: object is used as a proxy for objects that can be pickled
+# Note: mypy does not support cyclic definition currently
+Hdf5ConvertibleValues = Union[  # type: ignore
+    int, float, Batch, np.ndarray, torch.Tensor, object,
+    'Hdf5ConvertibleType',  # type: ignore
+]
+
+Hdf5ConvertibleType = Dict[str, Hdf5ConvertibleValues]  # type: ignore
+
+
+def to_hdf5(x: Hdf5ConvertibleType, y: h5py.Group) -> None:
+    """Copy object into HDF5 group."""
+    for k, v in x.items():
+        if isinstance(v, (Batch, dict)):
+            # dicts and batches are both represented by groups
+            subgrp = y.create_group(k)
+            if isinstance(v, Batch):
+                subgrp_data = v.__getstate__()
+            else:
+                subgrp_data = v
+            subgrp.attrs["__data_type__"] = v.__class__.__name__
+            to_hdf5(subgrp_data, subgrp)
+        elif isinstance(v, (np.ndarray, torch.Tensor)):
+            # NumPy arrays and PyTorch tensors are written to datasets
+            y.create_dataset(k, data=to_numpy(v))
+            y[k].attrs["__data_type__"] = v.__class__.__name__
+        elif isinstance(v, (int, float)):
+            # ints and floats are stored as attributes of groups
+            y.attrs[k] = v
+        else:  # resort to pickle for any other type of object
+            try:
+                int_data = np.frombuffer(pickle.dumps(v), dtype="uint8")
```

Comment:
I don't quite get why does it need "uint8" here.

Reply:
I used `frombuffer` because HDF5's handling of strings excludes strings with null bytes. However, the data type wasn't inferred correctly somehow. If I don't specify a data type I get `ValueError: buffer size must be a multiple of element size`. I can see that "uint8" is a confusing choice though. I have replaced it with `np.byte`.